core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -284,12 +289,18 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
                   }
                   // If the broker is a follower, updates the isr in ZK and notifies the current leader
                   replicaStateMachine.handleStateChanges(Set(PartitionAndReplica(topicAndPartition.topic,
-                    topicAndPartition.partition, id)), OfflineReplica)
+                    topicAndPartition.partition, id)), OfflineReplica, newBatch = false)
                 }
               }
             }
           }
       }
+
+      inLock(controllerContext.controllerLock) {
```

Comment:
I _think_ this works, but what we will lose with this is a safety check - i.e., an assertion that a state change has completed prior to handling another state change. (A complete state change being defined as update ZK and enqueue controller requests to be sent out to brokers). Ideally we should come up with a way to preserve that assertion to avoid the possibility of interleaved state changes.